### PR TITLE
pkgs/stdenv/generic/setup.sh: lint with ShellCheck

### PR DESCRIPTION
**Pull request by ShamrockLee**
The behaviour of `[[ -n/-z "${FOO[@]}" ]]` is unspecified. Use `[[ -n/-z "${FOO[*]-}" ]]` instead.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by ShamrockLee** [comment]
@ofborg build stdenv
@ofborg build hello hello.tests

File: pkgs/stdenv/generic/setup.sh [hunk 1]
```diff
@@ -396,7 +397,7 @@ concatTo() {
     for arg in "$@"; do
         IFS="=" read -r name default <<< "$arg"
         local -n nameref="$name"
-        if [[ ! -n "${nameref[@]}" && -n "$default" ]]; then
+        if [[ -z "${nameref[*]-}" && -n "${default-}" ]]; then
```
[the remark this answers is not in the document]
Comment (ShamrockLee) [reply pkgs/stdenv/generic/setup.sh:400]: They allow undefined variables to be treated as if their value were empty.

Maybe I should remove them, as it changes the implementation.

[the remark this answers is not in the document]
Comment (ShamrockLee) [reply pkgs/stdenv/generic/setup.sh:400]: I should remove the hyphens:

```bash
        if [[ -z "${nameref[*]}" && -n "$default" ]]; then
```

As the Bash Reference Manual doesn't specify the behavior of `[[ -n "${arr[@]}" ]]`, let's experiment with the command `if [[ -n "${foo[@]}" ]]; then echo yes; else echo no; fi; unset foo`:

| `unset foo` | `foo=""` | `foo=" "` | `foo=" hi"` | `foo="hello hi"` |
|-|-|-|-|-|
| no | no | yes | yes | yes |

|`foo=()`|`foo=("")`|`foo=("" "")` | `foo=("" hi)` | `foo=("hello" "hi")` |
|-|-|-|-|-|
| no | no | yes | yes | yes |

According to the Bash Reference Manual,

> Any element of an array may be referenced using ${name[subscript]}. ... If the word is double-quoted, ${name[*]} expands to a single word with the value of each array member separated by the first character of the IFS variable

This means the following conversion when the first character of the IFS variable is `" "`:

|`foo=()`|`foo=("")`|`foo=("" "")` | `foo=("" hi)` | `foo=("hello" "hi")` |
|-|-|-|-|-|
| `""` | `""` | `" "` | `" hi"` | `"hello hi"` |

It should be safe to say that `[[ -n "${foo[@]}" ]]` behaves like `[[ -n "$foo[*]" ]]`, but only the latter is specified in the Bash Reference Manual.

[the remark this answers is not in the document]
Comment (ShamrockLee) [reply pkgs/stdenv/generic/setup.sh]: Updated.

**Review by wolfgangwalther — APPROVED** [review]

**Pull request merged**